src/binder/query/query_graph_label_analyzer.cpp

```diff
@@ -93,49 +91,182 @@ void QueryGraphLabelAnalyzer::pruneNode(const QueryGraph& graph, NodeExpression&
+                backwardPrunedEntries.insert(backwardPrunedEntries.getEntryIndex(i - 1), i);
+            } else {
+                dstTableIDSet = collectRelNodes("<-", prunedEntries);
+                backwardPrunedEntries.insert(prunedEntries, i);
             }
-            prunedEntries.push_back(entry);
         }
     }
-    rel.setEntries(prunedEntries);
+    backwardPrunedEntries.reverse();
+    auto upperPE = PrunedEntries::merge(forwardPrunedEntries, backwardPrunedEntries);
+    if (lowerBound != upperBound) {
+        for (auto i = lowerBound; i < upperBound; ++i) {
+            auto forwardSubPrunedEntries = forwardPrunedEntries.sub(0, i);
+            auto backwardSubPrunedEntries = backwardPrunedEntries.sub(upperBound - i, upperBound);
+            auto tempPrunedEntries =
+                PrunedEntries::merge(forwardSubPrunedEntries, backwardSubPrunedEntries);
+            upperPE.unionPrunedEntries(tempPrunedEntries);
+        }
+    }
+    return upperPE;
+}
+
+std::vector<catalog::TableCatalogEntry*> QueryGraphLabelAnalyzer::getTableCatalogEntries(
+    std::vector<common::table_id_t> tableIDs) const {
+    std::vector<catalog::TableCatalogEntry*> relEntries;
+    for (const auto& tableID : tableIDs) {
+        relEntries.push_back(catalog->getTableCatalogEntry(tx, tableID));
+    }
+    return relEntries;
+}
+
+void QueryGraphLabelAnalyzer::pruneRel(RelExpression& rel) const {
+    auto srcTableIDSet = rel.getSrcNode()->getTableIDsSet();
+    auto dstTableIDSet = rel.getDstNode()->getTableIDsSet();
+    if (rel.isRecursive()) {
+        auto nodeTableIDs = catalog->getNodeTableIDs(tx);
+        // there is no label on both sides
+        if (srcTableIDSet.size() == dstTableIDSet.size() &&
+            dstTableIDSet.size() == nodeTableIDs.size()) {
+            return;
+        }
+
+        common::table_id_set_t allNodeTableIDs{nodeTableIDs.begin(), nodeTableIDs.end()};
+        auto prunedEntries = pruneRecursiveRel(rel.getEntries(), srcTableIDSet, dstTableIDSet,
+            rel.getDirectionType(), rel.getLowerBound(), rel.getUpperBound(), allNodeTableIDs);
+
+        if (!prunedEntries.empty()) {
+            std::vector<std::vector<table_id_t>> stepActivationRelInfos;
+            std::unordered_set<TableCatalogEntry*> temp;
+            for (const auto& item : prunedEntries.getPrunedEntries()) {
+                temp.insert(item.begin(), item.end());
+                std::vector<table_id_t> activationTabelIDs;
+                for (const auto& entry : item) {
+                    activationTabelIDs.push_back(entry->getTableID());
+                }
+                stepActivationRelInfos.push_back(activationTabelIDs);
+            }
+
+            rel.getRecursiveInfo()->stepActivationRelInfos = stepActivationRelInfos;
+            std::vector<catalog::TableCatalogEntry*> newRelEntries{temp.begin(), temp.end()};
+            if (!same(newRelEntries, rel.getEntries())) {
+                rel.setEntries(newRelEntries);
```

Comment:
As I understand you are trying prune src & dst node labels here. This is not the best place though.

Recall at the very top level of this rewriter we have

```
void QueryGraphLabelAnalyzer::pruneLabel(QueryGraph& graph) {
    for (auto i = 0u; i < graph.getNumQueryNodes(); ++i) {
        pruneNode(graph, *graph.getQueryNode(i));
    }
    for (auto i = 0u; i < graph.getNumQueryRels(); ++i) {
        pruneRel(*graph.getQueryRel(i));
    }
}

```

Ideally this should be an iterative process until coverage. So something like
```
while (!converged) {
      pruneNodes()
      pruneRels()
}
```

We don't do the iterative way because we don't wanna spend too much time in this rewrite stage. 

In your case, if this rewrite of node labels is important. I would suggest the following

```
pruneNodes()
pruneRels()
pruneNodes()
```

Reply:
Because only the src of the first hop and the dst of the last hop need to be adjusted, the logic is quite different from prunNode, so I have handled it here.